hexf/main.swift

```diff
@@ -0,0 +1,188 @@
+//
+//  main.swift
+//  hexf
+//
+//  Created as main.m by Kevin Wojniak on 9/24/17.
+//  Converted to main.swift by Reed Harston on 10/31/22.
+//  Copyright © 2017 ridiculous_fish. All rights reserved.
+//
+
+import Cocoa
+
+enum HexfError: Error {
+    case invalidUsage
+    case standardInputNoData
+    case launchAppNoUrl
+    case launchAppFailure
+}
+
+struct Controller {
+    private static let kAppIdentifier = "com.ridiculousfish.HexFiend"
+    
+    func printUsage() {
+        fputs("""
+Usage:
+
+  Open files:
+    hexf file1 [file2 file3 ...]
+
+  Compare files:
+    hexf -d file1 file2
+
+  Open piped data:
+    echo hello | hexf
+
+  Show help:
+    hexf -h | --help
+
+""", stderr)
+    }
+    
+    private static func standardize(path: String) -> String {
+        let url = URL(fileURLWithPath: path)
+        return url.path // get absolute path
+    }
+    
+    private var appRunning: Bool {
+        NSRunningApplication.runningApplications(withBundleIdentifier: Self.kAppIdentifier).first != nil
+    }
+    
+    private func launchApp(with args: [String]) throws {
+        guard let url = NSWorkspace.shared.urlForApplication(withBundleIdentifier: Self.kAppIdentifier) else {
+            fputs("Failed to get url to app bundle for: \(Self.kAppIdentifier)", stderr)
+            throw HexfError.launchAppNoUrl
+        }
+        
+        let config = [NSWorkspace.LaunchConfigurationKey.arguments: args]
+        
+        do {
+            // TODO: Heed deprecation warning, and get right config type. This will require an availability check.
```

Comment:
also fine with the newer option here behind an available check.

Reply:
I hit a hiccup with this one.

The new [openApplication method](https://developer.apple.com/documentation/appkit/nsworkspace/3172700-openapplication) uses a callback (or Swift 5.5's async await), so I can't directly throw if it gives an error.

This new code is invalid because you can't throw from the closure since the `launchApp` method call `openApplication` synchronously and moves on, returning before the closure completes and would throw the error. Right now the 
```swift
private func launchApp(with args: [String]) throws {
    ...
    let config = NSWorkspace.OpenConfiguration()
    config.arguments = args
    NSWorkspace.shared.openApplication(at: url, configuration: config) { (_, error) in
        if let error = error {
            fputs("Launch app failed: \(error.localizedDescription)", stderr)
            throw HexfError.launchAppFailure
        }
    }
}
```

Right now the only solutions I can see are to do something hacky to wait for the closure to finish, like in this [SO answer](https://stackoverflow.com/a/48606815), or to pass a completion handler to `launchApp` to handle the error at the top level. The completion handler is the more correct solution in my opinion, but it is a bit annoying that it would be required only for the macOS 10.15+ case and not the other.

Actually, as I typed that out I realized that is the better option since moving forward the 10.14 case is going to eventually be removed, leaving only the 10.15+ case and then everything will be right and clean. I'll go with that option then. (But I'm still open to other ideas.)